### PR TITLE
[CP] Extend CP to support load-balancing shards

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #133736
* #132820
* __->__ #132442

This PR extends the current ring attention to support load-balancing shards -- the context/sequence is divided into `2 * world_size` shards and each rank gets `rank` and `(world_size * 2 - rank - 1)` shards. The data re-shuffling is done in the `context_parallel` API.

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/distributed/_tensor/experimental/attention.py [hunk 1]
```diff
@@ -773,10 +852,72 @@ def attention_output_fn(mesh: DeviceMesh, outputs: Any) -> Any:
...
+
+    @classmethod
+    @abstractmethod
+    def unshard(
+        cls, buffer: torch.Tensor, mesh: DeviceMesh, seq_dim: int
+    ) -> torch.Tensor:
+        ...
+
+
+class EvenSharder(_LoadBalancer):
```
Comment (wanchaol) [inline torch/distributed/_tensor/experimental/attention.py:871]: shall we make this a private API too given that these are not user facing

File: torch/distributed/_tensor/experimental/attention.py [hunk 2]
```diff
@@ -773,10 +852,72 @@ def attention_output_fn(mesh: DeviceMesh, outputs: Any) -> Any:
...
+    def unshard(
+        cls, buffer: torch.Tensor, mesh: DeviceMesh, seq_dim: int
+    ) -> torch.Tensor:
+        buffer = buffer.contiguous()
+        all_buffers = [torch.empty_like(buffer) for _ in range(mesh.size())]
+        ft_c.all_gather_inplace(all_buffers, buffer, mesh)
+        return torch.cat(all_buffers, dim=seq_dim)
+
+
+class StripeLoadBalancer(_LoadBalancer):
```
Comment (wanchaol) [inline torch/distributed/_tensor/experimental/attention.py:889]: ditto: a private API if it's not user facing

File: torch/distributed/_tensor/experimental/attention.py [hunk 3]
```diff
@@ -857,3 +998,16 @@ def context_parallel(
         if original_buffer is not None:
             buffer.resize_(original_buffer.shape)
             buffer.copy_(original_buffer)
+
+
+@torch.no_grad()
+def context_parallel_unshard(
```
Comment (wanchaol) [inline torch/distributed/_tensor/experimental/attention.py:1004]: is this a conjugate function to `_context_parallel_buffers`? If so, maybe we should call it `_context_parallel_unshard_buffers`?

**Comment by fegin** [comment]
Add a TODO to track the propagation rule issue.

File: torch/distributed/_tensor/experimental/attention.py (hunk 3 above)
Comment (fegin) [reply torch/distributed/_tensor/experimental/attention.py:1004]: This is to unshard the output, which should be a public API. I think that's the only one we need to unshard from public usage perspective. Internally, we will use it for testing purpose. Since the public API is just `context_parallel`, I named this as `context_parallel_unshard`. Let me know if you have other preferences.

File: torch/distributed/_tensor/experimental/attention.py [hunk 4]
```diff
@@ -73,39 +83,70 @@ def _maybe_wait(tensor: torch.Tensor) -> torch.Tensor:
...
+            assert self._out is not None
+            lse = self._lse.chunk(2, dim=self._seq_dim)[1] if partial else self._lse
+            out = self._out.chunk(2, dim=self._seq_dim)[1] if partial else self._out
+
             # The algorithm from
             # github.com/zhuzilin/ring-flash-attention/pull/34#issuecomment-2076126795
             # gives a relatively stable result.
-            self._out = self._out - F.sigmoid(block_lse - self._lse) * (
-                self._out - block_out
+            new_lse = lse + torch.log(1 + torch.exp(block_lse - lse))
```
Comment (wconstab) [inline torch/distributed/_tensor/experimental/attention.py:129]: why this change?

File: torch/distributed/_tensor/experimental/attention.py [hunk 5]
```diff
@@ -243,7 +285,7 @@ def _templated_ring_attention(
     key = key.contiguous()
     value = value.contiguous()
 
-    sdpa_merger = _SDPAMerger(_convert_to_f32)
+    sdpa_merger = _SDPAMerger(_cp_options.convert_to_f32, seq_dim=2)
```
Comment (wconstab) [inline torch/distributed/_tensor/experimental/attention.py:288]: hardcoded seq_dim?

File: torch/distributed/_tensor/experimental/attention.py [hunk 6]
```diff
@@ -776,10 +863,72 @@ def attention_output_fn(mesh: DeviceMesh, outputs: Any) -> Any:
...
+    @classmethod
+    def shard(
+        cls, buffer: torch.Tensor, mesh: DeviceMesh, seq_dim: int
+    ) -> torch.Tensor:
+        cp_world_size = mesh.size()
+        cp_rank = mesh.get_local_rank()
+        assert buffer.size()[seq_dim] % (cp_world_size * 2) == 0
+        chunks = buffer.chunk(cp_world_size * 2, dim=seq_dim)
+        return torch.cat(
+            (chunks[cp_rank], chunks[cp_world_size * 2 - cp_rank - 1]),
```
Comment (wconstab) [inline torch/distributed/_tensor/experimental/attention.py:910]: is it good to hardcode that we only do 2 chunks for striping?  why not 4? (should this be adjustable, or is there a reason 2 is sufficient)

File: torch/distributed/_tensor/experimental/attention.py [hunk 7]
```diff
@@ -73,39 +83,70 @@ def _maybe_wait(tensor: torch.Tensor) -> torch.Tensor:
...
 
+def _partial_update(
+    original: torch.Tensor,
+    new: torch.Tensor,
+    dim: int,
+    n_chunks: int,
+    idx: int,
+    add: bool,
+) -> torch.Tensor:
+    chunks = list(original.chunk(n_chunks, dim=dim))
```
Comment (wconstab) [inline torch/distributed/_tensor/experimental/attention.py:94]: worth adding a docstring for the behavior.  and should there be an assert for the relative size of new and original?

File: torch/distributed/_tensor/experimental/attention.py [hunk 8]
```diff
@@ -73,39 +83,70 @@ def _maybe_wait(tensor: torch.Tensor) -> torch.Tensor:
...
-
-    def step(self, out: torch.Tensor, lse: torch.Tensor) -> None:
+            if partial:
+                self._lse = _partial_update(self._lse, new_lse, 2, 2, 1, add=False)
+                self._out = _partial_update(self._out, out, 2, 2, 1, add=False)
+            else:
+                self._lse = new_lse
+                self._out = out
+
+    def step(self, out: torch.Tensor, lse: torch.Tensor, partial: bool) -> None:
```
Comment (wconstab) [inline torch/distributed/_tensor/experimental/attention.py:141]: not for this PR,
i'm curious why we need to move lse to f32.  Should it be enough to move out to f32 (the first time we create out).  then we keep doing out += lse iiuc?  would it be ok to have f32 += bf16?

Comment (wconstab) [reply torch/distributed/_tensor/experimental/attention.py:141]: oh nevermind. we don't accumulate lse into out. we accumulate partials into both of them.

File: torch/distributed/_tensor/experimental/attention.py (hunk 4 above)
Comment (fegin) [reply torch/distributed/_tensor/experimental/attention.py:129]: good catch, this is a merge conflict.

File: torch/distributed/_tensor/experimental/attention.py (hunk 6 above)
Comment (fegin) [reply torch/distributed/_tensor/experimental/attention.py:910]: 2 should be sufficient and can reduce the ring attention implementation complexity (also less memory reshuffling). If we really think more than 2 is necessary, we can change it later.

Comment (wconstab) [reply torch/distributed/_tensor/experimental/attention.py:952]: i see. maybe at least make it a named constant so its clear which 2's are related to this and which ones arent?

File: torch/distributed/_tensor/experimental/attention.py [hunk 9]
```diff
@@ -73,39 +83,76 @@ def _maybe_wait(tensor: torch.Tensor) -> torch.Tensor:
...
+    dim: int,
+    n_chunks: int,
+    idx: int,
+    add: bool,
+) -> torch.Tensor:
+    """
+    This API partially update a chunk of ``original`` tensor. The ``original``
+    tensor will be first chunked along ``dim`` dimension then the ``idx`` chunk
+    will be updated with ``new``. If ``add`` is True, the chunk will be added
+    with ``new``, otherwise the chunk with be replaced by ``add``.
```
Comment (wconstab) [inline torch/distributed/_tensor/experimental/attention.py:98]: replaced by ``new`` right?

File: torch/distributed/_tensor/experimental/attention.py [hunk 10]
```diff
@@ -392,6 +467,7 @@ def _templated_ring_attention_backward(
             pointer += value.numel()
 
         if i != size - 1:
+            # Send the kv to the next rank.
```
Comment (wconstab) [inline torch/distributed/_tensor/experimental/attention.py:470]: so is world_size for CP typically 2, and that matches up to 2 chunks and 2 rotation steps? or am i confusing stripe-chunks with algorithm steps

File: torch/distributed/_tensor/experimental/attention.py [hunk 11]
```diff
@@ -400,13 +476,43 @@ def _templated_ring_attention_backward(
...
+            if i == 0 or (not _cp_options.enable_load_balance or not is_causal):
+                # We need to do SDPA with the full local q, k, v.
+                q, k, v, out_, dout, lse = (query, key, value, out, grad_out, logsumexp)
+            elif i <= rank:
+                # Striped load balancing case, and i <= rank.
+                # We need to do SPDA with only the first half of the k, v.
+                # Note that q, k, v, each contains two stripes.
+                q, k, v, out_, dout, lse = (
+                    query,
+                    key.chunk(2, dim=2)[0],
```
Comment (wconstab) [inline torch/distributed/_tensor/experimental/attention.py:488]: i think you need to name the constants for readability. and for assertability.  We probably should be asserting that ` CP == 2 (if enabled) since various things assume this.  

For example the `elif i <= rank` and then `else` branches seem to be an assumption of CP=2 iiuc?

File: torch/distributed/_tensor/experimental/attention.py [hunk 12]
```diff
@@ -243,37 +294,59 @@ def _templated_ring_attention(
...
-
-            sdpa_merger.step(out, logsumexp)
+        else:
+            # Striped load balancing case, and i > rank.
+            # We need to do SPDA with only the second half of the q, and update
+            # only the the second part of  logsumexp. So partial is True.
+            # Note that q, k, v, each contains two stripes.
+            q, k, v, partial = query.chunk(2, dim=2)[1], key, value, True
+
+        out, logsumexp, *rest = op(
```
Comment (wconstab) [inline torch/distributed/_tensor/experimental/attention.py:342]: it is actually not easy to trace through the math, becuase it is not clear what 'op' is doing. do you know if the various SDPA kernels have some docstring somewhere? i did not find documentation for the _ring or _efficient kernels.  (only for the regular SDPA kernel, which does not do the same thing as these other kernels)

Comment (fegin) [reply torch/distributed/_tensor/experimental/attention.py:342]: Add a comment.

File: torch/distributed/_tensor/experimental/attention.py [hunk 13]
```diff
@@ -936,17 +992,23 @@ def unshard(
...
 
-class _StripeLoadBalancer(_LoadBalancer):
+class _RoundRobinLoadBalancer(_LoadBalancer):
     """
-    This load balancer chunk the buffer into cp_world_size * 2 shards, and
-    rank0 gets 0th and (cp_world_size * 2 - 1)th shards; aank1 will get 1st
-    and (cp_world_size * 2 - 2)th shards; ...
+    This load balancer chunk the buffer into cp_world_size * ROUND_ROBIN_CYCLE
+    shards, and uses a round robine approach to achieve load balancing.
+    Since ROUND_ROBIN_CYCLE begin 2 is enough for the causal masking, we currently
```
Comment (wconstab) [inline torch/distributed/_tensor/experimental/attention.py:999]: typos (robine, begin 2)

clarity: not sure why "2 is enough for causal masking"?

File: torch/distributed/_tensor/experimental/attention.py [hunk 14]
```diff
@@ -246,37 +322,61 @@ def _templated_ring_attention(
...
+            # Round-robin load balancing case, and i <= rank.
+            # We need to do SPDA, with only the first half of the k, v.
+            # Note that q, k, v, each contains two chunks.
+            q, k, v, partial = (
                 query,
-                key,
-                value,
-                is_causal=is_causal_behavior.value,
-                **kwargs,
+                key.chunk(2, dim=2)[0],
```
Comment (wconstab) [inline torch/distributed/_tensor/experimental/attention.py:359]: missed updating '2' constant here

File: torch/distributed/_tensor/experimental/attention.py [hunk 15]
```diff
@@ -246,37 +322,61 @@ def _templated_ring_attention(
...
             rank=rank, world_size=size, i=i, is_causal=is_causal
         )
 
-        if is_causal_behavior != _CausalBehavior.SKIP:
-            out, logsumexp, *rest = op(
+        if is_causal_behavior == _CausalBehavior.SKIP:
+            continue
+
+        if i == 0 or (not _cp_options.enable_load_balance or not is_causal):
+            # We need to do SDPA with the full local q, k, v.
```
Comment (wconstab) [inline torch/distributed/_tensor/experimental/attention.py:351]: we discussed offline about adding a more detailed 'picture' of the round-robin splitting of QKV and using that to help explain the if/elif/else.

But one minor change  that would help, "full local" can be replaced with "both local chunks of" to avoid ambiguity

File: torch/distributed/_tensor/experimental/attention.py [hunk 16]
```diff
@@ -285,6 +285,103 @@ def _templated_ring_attention(
...
+    In the second iteraion, rank0 owns (q0, q3) and (k1, k2); rank1 owns (q1, q2) and
+    (k0, k3). For the local query, q0, rank0 doesn't have to any computation. But
+    both the result of q3k1 and q3k2 are required. So we need to chunk out q3 as that
+    is the only query needed to call SDPA. This corresponds to the `else` in the
+    if/elif/else statement in the implementation. As for rank1, the key, k0, is not
+    required for q1 and q2. So we need to chunk out k3 as it is the only key required
+    for calling SDPA. This corresponds to the `elif` statement.
+
+
+    Algorithm Description:
```
Comment (wconstab) [inline torch/distributed/_tensor/experimental/attention.py:334]: accidentally pasted the description twice?

File: torch/distributed/_tensor/experimental/attention.py [hunk 17]
```diff
@@ -285,6 +285,103 @@ def _templated_ring_attention(
     """
     This is a generalized ring attention implementation that can support multiple attention ops.
 
+    Algorithm:
```
Comment (wconstab) [inline torch/distributed/_tensor/experimental/attention.py:288]: better to use 'Note' format from other pytorch docs.  then you can refer to this note by mentioning 'see note ... ' in some other code comments.

e.g.
aten/src/ATen/CPUGeneratorImpl.cpp: ` * See Note [Acquire lock when using random generators]`

aten/src/ATen/core/Generator.h: `* Note [Acquire lock when using random generators]`

**Review by wconstab — APPROVED** [review]
the explanation is great. i appreciate the work! pls fix the paste issues and consider the 'Note' format

**Comment by fegin** [comment]
@pytrochbot merge -f "The failing tests are not related."

**Comment by rishub-tamirisa** [comment]
Hello, thanks for implementing this, I've been running the experimental ring attn implementation for long-context SFT the last few days and it works great! 

I wanted to confirm the preferred way to run completion-only fine-tuning with the following setup. I'm using `context_parallel` on the same mesh dim as FSDP across 4 devices, and the non-ignored labels (labels that are not set to -100 to ignore during completion-only SFT) only exist on the last device, since I don't want to backprop on the majority of the context. 

Naively doing this results in a `NaN` loss on all ranks except the last, which is expected; if FSDP already ignores `NaN` losses then this might already be handled correctly under the hood, but is it possible that things are going wrong silently?

cc @awgu @fegin @d4l3k
